proxy/coco/src/state.rs

```diff
@@ -0,0 +1,733 @@
+//! Utility to work with the peer api of librad.
+
+use std::convert::TryFrom;
+use std::future::Future;
+use std::net::{IpAddr, SocketAddr};
+use std::path::{self, PathBuf};
+use std::sync::Arc;
+use std::time::Duration;
+
+use tokio::sync::Mutex;
+
+use librad::git::local::{transport, url::LocalUrl};
+use librad::git::refs::Refs;
+use librad::git::storage;
+use librad::keys;
+use librad::meta::entity;
+use librad::meta::project as librad_project;
+use librad::meta::user;
+use librad::net::gossip::PeerInfo;
+use librad::net::peer::PeerApi;
+use librad::paths;
+use librad::peer::PeerId;
+use librad::uri::{RadUrl, RadUrn};
+use radicle_keystore::sign::Signer as _;
+use radicle_surf::vcs::git;
+use radicle_surf::vcs::git::git2;
+
+use crate::error::Error;
+use crate::project;
+use crate::seed::Seed;
+use crate::signer;
+use crate::user::{verify as verify_user, User};
+
+/// Thread-safe wrapper for [`State`].
+pub type Lock = Arc<Mutex<State>>;
+
+impl From<State> for Lock {
+    fn from(state: State) -> Self {
+        Arc::new(Mutex::new(state))
+    }
+}
+
+/// High-level interface to the coco monorepo and gossip layer.
+pub struct State {
+    /// Internal handle on [`PeerApi`].
+    pub(crate) api: PeerApi<keys::SecretKey>,
+}
+
+impl State {
+    /// Create a new [`State`] given a [`PeerApi`].
+    #[must_use]
+    pub fn new(api: PeerApi<keys::SecretKey>, signer: signer::BoxedSigner) -> Self {
```

Comment:
I prefer the former, because the qualified use gives you context where it originates from. Esp. in a codebase like this where a lot of the sub-domains reappear and are reused.